cmd/runner/main.go

```diff
@@ -1,56 +1,76 @@
+	"github.com/benchttp/runner/configparser"
 	"github.com/benchttp/runner/request"
 )
 
-const (
-	DefaultConcurrency = 1
-	DefaultRequests    = 0 // Use duration as exit condition if omitted.
-	DefaultDuration    = 60 * time.Second
-	DefaultTimeout     = 10 * time.Second
-)
-
 var (
-	url         string
-	concurrency int           // Number of connections to run concurrently.
-	requests    int           // Number of requests to run, use duration as exit condition if omitted.
-	duration    time.Duration // Duration of test, in seconds.
-	timeout     time.Duration // Timeout for each http request, in seconds.
+	configFile    string
+	url           string
+	concurrency   int           // Number of connections to run concurrently
+	requests      int           // Number of requests to run, use duration as exit condition if omitted.
+	timeout       time.Duration // Timeout for each http request
+	globalTimeout time.Duration // Duration of test
 )
 
-func parseArgs() {
-	url = os.Args[len(os.Args)-1]
+var defaultConfigFiles = []string{
+	"./.benchttp.yml",
+	"./.benchttp.yaml",
+	"./.benchttp.json",
+}
 
-	flag.IntVar(&concurrency, "c", DefaultConcurrency, "Number of connections to run concurrently")
-	flag.IntVar(&requests, "r", DefaultRequests, "Number of requests to run, use duration as exit condition if omitted")
-	flag.DurationVar(&duration, "d", DefaultDuration, "Duration of test, in seconds")
-	flag.DurationVar(&timeout, "t", DefaultTimeout, "Timeout for each http request, in seconds")
+func parseArgs() {
+	flag.StringVar(&configFile, "configFile", findFile(defaultConfigFiles), "Config file path")
+	flag.StringVar(&url, "url", "", "Target URL to request")
+	flag.IntVar(&concurrency, "concurrency", 0, "Number of connections to run concurrently")
+	flag.IntVar(&requests, "requests", 0, "Number of requests to run, use duration as exit condition if omitted")
+	flag.DurationVar(&timeout, "timeout", 0, "Timeout for each http request")
+	flag.DurationVar(&globalTimeout, "globalTimeout", 0, "Duration of test")
 	flag.Parse()
-
-	fmt.Printf("Testing url: %s\n", url)
-	fmt.Printf("concurrency: %d\n", concurrency)
-	if requests > 0 {
-		fmt.Printf("requests: %d\n", requests)
-	}
-	fmt.Printf("duration: %s\n", duration)
-
-	println()
 }
 
 func main() {
 	parseArgs()
 
-	ctx, cancel := context.WithTimeout(context.Background(), duration)
-	defer cancel()
+	cfg := makeRunnerConfig()
+	fmt.Println(cfg)
 
-	rec := request.Do(ctx, requests, concurrency, url, timeout)
+	rec := request.Do(cfg)
 
 	fmt.Println("total:", len(rec))
 }
+
+// makeRunnerConfig retrieves a config from
+func makeRunnerConfig() config.Config {
+	fileConfig, err := configparser.Parse(configFile)
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		// config file is not mandatory, other errors are critical
+		log.Fatal(err)
+	}
+
+	cliConfig := config.New(url, requests, concurrency, timeout, globalTimeout)
+
+	cfg := config.Merge(fileConfig, cliConfig)
+
+	return cfg
+}
```

Comment:
This is not a very pleasing api. It works but we can do better ! Suggestion: 

How I want to use it as a main ?

```go
func main() {
	cfg, err := config.Parse(file)
	if err != nil && !errors.Is(err, os.ErrNotExist) {
		log.Fatalf("could not configure the runner: %s", err)
	}

	cfg = config.Merge(cfg, args)
	// ...
}
```

We can even go further, as main, I don't really care about the parsing and merging steps. So we can have an api like :
```go
// package config

func Parse(file string) (Config, error)

func Merge(cfg Config, args Type?) (Config, error)

func ParseAndMerge(file string, args Type?) (Config, error) {
	return Merge(Parse(file), args)
}
```
Only `ParseAndMerge` could be public if it is the usecase in the end.

This obviously means packages `config` and `configparser` should be merge. I'll have another comment about that.

Reply:
Strongly disagree: `Config` has nothing to do with reading and parsing files and should know nothing about that business. When `request` imports `config`, it should not import a parsing module.

Reply:
I see your point, but if we look at what the purpose of each package we have this:

I use `configparser.Parse` to get a `config.Config`. I use `config.Parse` to get a `config.Config` and `config.Merge` to get another `config.Config`. 
These are the only functionalities offered by the 2 packages.

`request` import `config` to simply type the input of `request.Do`, it does not require any functionality.

`config` really just bundles params into a nice structure for use by other parts of the system. So I don't really think there is distinction between the reading and parsing config file and structuration of the config object.

Reply:
`Config` defines the behaviour of the whole runner. It's a domain entity that could possibly be used anywhere in the app. It's in `configparser`, it's in `request`, it'll also be in `output` (whatever the name) because we want to post that config as well.
`configparser` provides low-level utility to get a `Config` out of a file. `Config` must not be coupled with that.

I think we both made our points, however if anyone else has an opinion on that topic, they're welcome.

Reply:
I do not agree, it is not an _entity_ if it does not have behaviours attached. There is no `Config.Do` procedure anywhere. It's just a bag of parameters for other entities/actors (the requester, the thing that will send the results, ...).

All that is exported form `config` is its struct and `New` (or `NewButMerge`).
What it really is compared to an entity, is a simple value object. Same as `record` in fact.

Reply:
Side note: no thumbs down. It's an unconstructive way to express disagreement and should be avoided.

Reply:
When in doubt, I would tend to separate the functionalities in different packages.
I feel like even if we realize later that it is not ideal, it will be easier to regroup different functionalities in the same package than the opposite.